File: src/it/java/quebec/virtualite/backend/services/rest/RestServerSteps.java
package quebec.virtualite.backend.services.rest;

import cucumber.api.java.Before;
import cucumber.api.java.en.Given;
import cucumber.api.java.en.Then;
import cucumber.api.java.en.When;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.ContextConfiguration;
import quebec.virtualite.backend.services.domain.DomainService;
import quebec.virtualite.backend.utils.RestClient;

import javax.annotation.PostConstruct;

import static org.apache.http.HttpStatus.SC_OK;
import static org.hamcrest.core.Is.is;
import static org.junit.Assert.assertThat;
import static org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT;
import static quebec.virtualite.backend.security.SecurityUsers.TEST_PASSWORD;
import static quebec.virtualite.backend.security.SecurityUsers.TEST_USER;
import static quebec.virtualite.backend.utils.RestParam.param;

@SpringBootTest(webEnvironment = RANDOM_PORT)
@ContextConfiguration
public class RestServerSteps
{
    @Autowired
    private DomainService domainService;

    @Autowired
    private RestClient rest;

    @Value("${local.server.port}")
    private int serverPort;

    @PostConstruct
    public void _init()
    {
        rest._init(serverPort);
    }

    @Before
    public void beforeEachScenario()
    {
        domainService.deleteGreetings();
    }

    @Given("^we are logged in$")
    public void weAreLoggedIn()
    {
        rest.login(TEST_USER, TEST_PASSWORD);
    }

    @Given("^we are not logged in$")
    public void weAreNotLoggedIn()
    {
        // Nothing to do here
    }

    /**
     * Server Unit Test: {@link RestServerTest#greet()}
     */
    @When("we ask for a greeting for {string} [GET {string}]")
    public void weAskForAGreetingForGET(String name, String url)
    {
        rest.get(url, param("name", name));
    }

    @Then("^we get a greeting message$")
    public void weGetAGreetingMessage(String expectedJson)
    {
        assertThat(rest.response().statusCode(), is(SC_OK));
        assertThat(rest.response().asString(), is(rest.trim(expectedJson)));
    }

    @Then("we should get a {int} error")
    public void weShouldGetAError(int errorCode)
    {
        assertThat(rest.response().statusCode(), is(errorCode));
    }
}

File: src/it/java/quebec/virtualite/backend/services/rest/RestServerSteps.kt
package quebec.virtualite.backend.services.rest

import io.cucumber.java.Before
import io.cucumber.java.en.Given
import io.cucumber.java.en.Then
import io.cucumber.java.en.When
import org.apache.http.HttpStatus
import org.hamcrest.core.IsEqual.equalTo
import org.junit.Assert
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Value
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment
import org.springframework.test.context.ContextConfiguration
import quebec.virtualite.backend.security.SecurityUsers
import quebec.virtualite.backend.services.domain.DomainService
import quebec.virtualite.backend.utils.RestClient
import quebec.virtualite.backend.utils.RestParam
import javax.annotation.PostConstruct

@SpringBootTest(webEnvironment = WebEnvironment.RANDOM_PORT)
@ContextConfiguration
class RestServerSteps
{
    @Autowired
    private lateinit var domainService: DomainService

    @Autowired
    private lateinit var rest: RestClient

    @Value("\${local.server.port}")
    private val serverPort = 0

    @PostConstruct
    fun init()
    {
        rest._init(serverPort)
    }

    @Before
    fun beforeEachScenario()
    {
        domainService.deleteGreetings()
    }

    @Given("^we are logged in$")
    fun weAreLoggedIn()
    {
        rest.login(SecurityUsers.TEST_USER, SecurityUsers.TEST_PASSWORD)
    }

    @Given("^we are not logged in$")
    fun weAreNotLoggedIn()
    {
        // Nothing to do here
    }

    /**
     * Server Unit Test: [RestServerTest.greet]
     */
    @When("we ask for a greeting for {string} [PUT {string}]")
    fun weAskForAGreetingForGET(name: String, url: String)
    {
        rest.put(url, RestParam("name", name))
    }

    @Then("^we get a greeting message$")
    fun weGetAGreetingMessage(expectedJson: String)
    {
        Assert.assertThat(rest.response().statusCode(), equalTo(HttpStatus.SC_OK))
        Assert.assertThat(rest.response().asString(), equalTo(rest.trim(expectedJson)))
    }

    @Then("we should get a {int} error")
    fun weShouldGetAError(errorCode: Int)
    {
        Assert.assertThat(rest.response().statusCode(), equalTo(errorCode))
    }
}